File: elixir/anagram/anagram.exs
defmodule Anagram do

  def match(word, word_list) do
    word_list |> Enum.filter(fn(candidate) -> anagram?(word, candidate) end)
  end

  defp anagram?(word, candidate) do
    word != candidate && normalize(word) == normalize(candidate)
  end

  defp normalize(word) do 
    word 
      |> String.downcase 
      |> String.to_char_list! 
      |> Enum.sort 
  end
end

File: java/anagram/src/main/java/Anagram.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;

public class Anagram {

    private String letters;
    private String original;

    public Anagram (String string) {
        string = string.toLowerCase();
        char[] letters = string.toCharArray();
        Arrays.sort(letters);
        this.letters = new String(letters);
        this.original = string;
    }

    public ArrayList<String> match(List<String> words) {
        ArrayList<String> matches = new ArrayList<String>();

        words.forEach((word) -> {
            String lowerWord = word.toLowerCase();
            char[] letters = lowerWord.toCharArray();
            Arrays.sort(letters);
            if (!this.original.equals(lowerWord) && this.letters.equals(new String(letters))) {
                matches.add(word);
            }
        });

        return matches;
    }
}
